File: languages/python/bloock/entity/signature.py
from __future__ import annotations
from enum import Enum
import bloock._bridge.proto.authenticity_entities_pb2 as proto


class SignatureAlg(Enum):
    ECDSA = 0
    ENS = 1
    UNRECOGNIZED = -1

    def __int__(self):
        return self.value

    @staticmethod
    def from_str(alg: str) -> SignatureAlg:
        if alg == "ES256K":
            return SignatureAlg.ECDSA
        elif alg == "ENS":
            return SignatureAlg.ENS
        else:
            return SignatureAlg.UNRECOGNIZED


class Signature:
    def __init__(
        self, message_hash: str, signature: str, protected: str, header: SignatureHeader
    ) -> None:
        self.signature = signature
        self.protected = protected
        self.header = header
        self.message_hash = message_hash

    @staticmethod
    def from_proto(signature: proto.Signature) -> Signature:
        return Signature(
            signature=signature.signature,
            protected=signature.protected,
            header=SignatureHeader.from_proto(signature.header),
            message_hash=signature.message_hash,
        )

    def to_proto(self) -> proto.Signature:
        return proto.Signature(
            signature=self.signature,
            protected=self.protected,
            header=self.header.to_proto(),
            message_hash=self.message_hash,
        )

    def get_alg(self) -> SignatureAlg:
        return SignatureAlg.from_str(self.header.alg)


class SignatureHeader:
    def __init__(self, alg: str, kid: str) -> None:
        self.alg = alg
        self.kid = kid

    @staticmethod
    def from_proto(header: proto.SignatureHeader) -> SignatureHeader:
        return SignatureHeader(alg=header.alg, kid=header.kid)

    def to_proto(self) -> proto.SignatureHeader:
        return proto.SignatureHeader(alg=self.alg, kid=self.kid)

File: languages/java/src/main/java/com/bloock/sdk/entity/authenticity/Signature.java
package com.bloock.sdk.entity.authenticity;

import com.bloock.sdk.bridge.proto.BloockAuthenticityEntities;

/**
 * Represents a cryptographic signature along with additional metadata.
 */
public class Signature {
  /**
   * Is the cryptographic signature.
   */
  String signature;
  /**
   * Is the algorithm used for the signature.
   */
  String alg;
  /**
   * Is the key identifier associated with the signature. (public key or key ID).
   */
  String kid;
  /**
   * Is the hash of the message that was signed.
   */
  String messageHash;
  /**
   * Is an optional field representing the subject of the signature.
   */
  String subject;

  /**
   * Constructs a Signature object with the specified parameters.
   * 
   * @param signature
   * @param alg
   * @param kid
   * @param messageHash
   * @param subject
   */
  Signature(String signature, String alg, String kid, String messageHash, String subject) {
    this.signature = signature;
    this.alg = alg;
    this.kid = kid;
    this.messageHash = messageHash;
    this.subject = subject;
  }

  public static Signature fromProto(BloockAuthenticityEntities.Signature signature) {
    return new Signature(
        signature.getSignature(),
        signature.getAlg(),
        signature.getKid(),
        signature.getMessageHash(),
        signature.getSubject());
  }

  public BloockAuthenticityEntities.Signature toProto() {
    return BloockAuthenticityEntities.Signature.newBuilder()
        .setSignature(this.signature)
        .setAlg(this.alg)
        .setKid(this.kid)
        .setMessageHash(this.messageHash)
        .setSubject(this.subject)
        .build();
  }

  /**
   * Gets the cryptographic signature.
   * 
   * @return
   */
  public String getSignature() {
    return signature;
  }

  /**
   * Sets the cryptographic signature.
   * 
   * @param signature
   */
  public void setSignature(String signature) {
    this.signature = signature;
  }

  /**
   * Returns the SignatureAlg based on the algorithm specified in the Alg field.
   * 
   * @return
   */
  public SignatureAlg getAlg() {
    return SignatureAlg.fromString(this.alg);
  }

  /**
   * Gets the key identifier associated with the signature. (public key or key
   * ID).
   * 
   * @return
   */
  public String getKid() {
    return kid;
  }

  /**
   * Sets the hash of the message that was signed.
   * 
   * @param hash
   */
  public void setMessageHash(String hash) {
    this.messageHash = hash;
  }

  /**
   * Gets the subject of the signature.
   * 
   * @return
   */
  public String getSubject() {
    return subject;
  }

  /**
   * Sets the subject of the signature.
   * 
   * @param subject
   */
  public void setSubject(String subject) {
    this.subject = subject;
  }
}
